index.bs

```diff
@@ -4091,6 +4104,138 @@ The following extended attributes are applicable to [=iterable declarations=]:
 </pre>
 
 
+<h4 id="idl-async-iterable">Asynchronously iterable declarations</h4>
+
+An [=interface=] can be declared to be asynchronously iterable by using an
+<dfn id="dfn-async-iterable-declaration" export>asynchronously iterable declaration</dfn>
+(matching <emu-nt><a href="#prod-AsyncIterable">AsyncIterable</a></emu-nt>) in the body of the
+[=interface=].
+
+<pre class="syntax">
+    interface interface_identifier {
+      async_iterable&lt;key_type, value_type&gt;;
+    };
+</pre>
+
+Objects that [=implement=] an [=interface=] that is declared to be asynchronously iterable support
+being iterated over asynchronously to obtain a sequence of values.
+
+Note: In the ECMAScript language binding, an interface that is asynchronously iterable will have
+<code class="idl">entries</code>, <code class="idl">keys</code>, <code class="idl">values</code>,
+and {{@@asyncIterator}} properties on its [=interface prototype object=].
+
+Prose accompanying an [=interface=] with an [=asynchronously iterable declaration=] must define a
+<dfn id="dfn-get-the-next-iteration-result">get the next iteration result</dfn> algorithm.
+This algorithm receives a <b>[=this=]</b> value, which is an instance of the [=interface=] that it
+is defined for, and the <dfn export>current state</dfn>.
+It must return a {{Promise}} that either resolves with undefined – to signal the end of the
+iteration – or a tuple with three elements:
+
+1.  a value of the first type given in the declaration;
+1.  a value of the second type given in the declaration;
+1.  an opaque value that is passed back to the next invocation of the algorithm,
```

Comment:
Hmm. This wasn't what was discussed in https://github.com/heycam/webidl/issues/580. I prefer the formulation over there, if you think it's possible to spec.

Reply:
With "yield"? I added a commit to https://github.com/WICG/kv-storage/pull/68 to see what that looks like. It seems kinda weird to be throwing and yielding out of promise handlers and event listeners, but if you think it makes sense, I can try to draft prose for IDL to make it well-defined.

Reply:
I'll need to think on this harder. It does seem like we have two rather different ways spec authors could do this, one of which makes sense when promises are involved, one of which makes sense when background threads are involved. Thanks for the added commit to give something concrete for us to look at.

My guess, before thinking hard, is that we'll either:

- Allow both styles, at the spec author's option; or
- Start with what kv-storage wants, then expand to what file system wants, etc., figuring it out as we go.

Reply:
Let's stick with what you've got. Mixing yield/throw and promises is too weird, indeed.